pupil_src/shared_modules/marker_tracker_3d/worker/visibility_graphs.py

```diff
@@ -0,0 +1,147 @@
+            if self._n_frames_passed >= self._select_novel_markers_interval:
+                self._n_frames_passed = 0
+                self._save_current_camera_extrinsics(current_camera_extrinsics)
+                novel_markers = self._pick_novel_markers(marker_id_to_detections)
+                self._add_novel_markers_to_model_storage(novel_markers)
+
+        self._model_storage.current_frame_id += 1
+        self._n_frames_passed += 1
+
+    def _save_current_camera_extrinsics(self, current_camera_extrinsics):
+        if current_camera_extrinsics is not None:
+            self._model_storage.frame_id_to_extrinsics_opt[
+                self._model_storage.current_frame_id
+            ] = current_camera_extrinsics
+
+    def _pick_novel_markers(self, marker_id_to_detections):
+        if len(marker_id_to_detections) < self._min_n_markers_per_frame:
+            return []
+
+        novel_marker_candidates = self._get_novel_marker_candidates(
+            marker_id_to_detections
+        )
+
+        # if there are markers which have not yet been optimized,
+        # add all markers in this frame and
+        # do not need to check if the corresponding bins are available
+        if not bool(
+            marker_id_to_detections.keys()
+            - self._model_storage.marker_id_to_extrinsics_opt.keys()
+        ):
+            novel_marker_candidates = self._filter_novel_markers_by_bins_availability(
+                novel_marker_candidates
+            )
+
+        if len(novel_marker_candidates) < self._min_n_markers_per_frame:
+            return []
+
+        return novel_marker_candidates
+
+    def _get_novel_marker_candidates(self, marker_id_to_detections):
+        bins_x, bins_y = self._get_bins(marker_id_to_detections)
+        novel_marker_candidates = [
+            NovelMarker(
+                self._model_storage.current_frame_id,
+                marker_id,
+                marker_id_to_detections[marker_id]["verts"],
+                (x, y),
+            )
+            for marker_id, x, y in zip(marker_id_to_detections.keys(), bins_x, bins_y)
+        ]
+        return novel_marker_candidates
+
+    def _get_bins(self, marker_id_to_detections):
+        centroids = np.array(
+            [
+                marker_id_to_detections[k]["centroid"]
+                for k in marker_id_to_detections.keys()
+            ]
+        )
+        bins_x = np.digitize(centroids[:, 0], self._bins_x)
+        bins_y = np.digitize(centroids[:, 1], self._bins_y)
+        return bins_x, bins_y
+
+    def _filter_novel_markers_by_bins_availability(self, novel_marker_candidates):
+        novel_markers = []
+        for candidate in novel_marker_candidates:
+            n_same_markers_in_bin = len(
+                [
+                    marker
+                    for marker in self._model_storage.all_novel_markers
+                    if marker.marker_id == candidate.marker_id
+                    and marker.bin == candidate.bin
```

Comment:
Maybe I am missing something, but in terms of computational performance it would be more efficient to store an array that stores the list of contained markers in every bin. Then we could iterate through the bin markers only rather than iterating through all markers. Am I right in thinking that `all_novel_markers` are all markers in the key frames? Not sure why the `novel` is included!